crates/adapters/src/controller.rs

```diff
@@ -2981,6 +3034,186 @@ struct ControllerInit {
+        .collect::<BTreeMap<_, _>>();
+
+    old_configured_inputs.remove("now");
+
+    let old_configured_outputs = old_config
+        .outputs
+        .iter()
+        .filter(|(_, cfg)| !cfg.connector_config.transport.is_transient())
+        .map(|(name, cfg)| (name.clone(), cfg.clone()))
+        .collect::<BTreeMap<_, _>>();
+
+    let added_input_connectors = new_config
+        .inputs
+        .keys()
+        .filter(|k| !old_configured_inputs.contains_key(*k))
+        .map(|k| k.to_string())
+        .collect::<Vec<_>>();
+
+    let removed_input_connectors = old_configured_inputs
+        .iter()
+        .filter(|(k, config)| {
+            !new_config.inputs.contains_key(*k)
+                && !mir_diff
+                    .as_ref()
+                    .map(|mir_diff| mir_diff.removed_tables.contains(&config.stream.to_string()))
+                    .unwrap_or(true)
+        })
+        .map(|(k, _)| k.to_string())
+        .collect::<Vec<_>>();
+
+    let modified_input_connectors = new_config
+        .inputs
+        .iter()
+        .filter(|(k, v)| {
+            old_configured_inputs.contains_key(*k)
+                && !old_configured_inputs
+                    .get(*k)
+                    .unwrap()
+                    .connector_config
+                    .equal_modulo_paused(&v.connector_config)
+        })
+        .map(|(k, _)| k.to_string())
+        .collect::<Vec<_>>();
+
+    let added_output_connectors = new_config
+        .outputs
+        .keys()
+        .filter(|k| !old_configured_outputs.contains_key(*k))
+        .map(|k| k.to_string())
+        .collect::<Vec<_>>();
+
+    let removed_output_connectors = old_configured_outputs
+        .iter()
+        .filter(|(k, config)| {
+            !new_config.outputs.contains_key(*k)
+                && !mir_diff
+                    .as_ref()
+                    .map(|mir_diff| mir_diff.removed_views.contains(&config.stream.to_string()))
+                    .unwrap_or(true)
+        })
+        .map(|(k, _)| k.to_string())
+        .collect::<Vec<_>>();
+
+    let modified_output_connectors = new_config
+        .outputs
+        .iter()
+        .filter(|(k, v)| {
+            old_configured_outputs.contains_key(*k)
+                && !old_configured_outputs
+                    .get(*k)
+                    .unwrap()
+                    .connector_config
+                    .equal_modulo_paused(&v.connector_config)
+        })
+        .map(|(k, _)| k.to_string())
+        .collect::<Vec<_>>();
+
+    Ok(PipelineDiff {
+        program_diff: mir_diff.as_ref().ok().cloned(),
+        program_diff_error: mir_diff.err(),
+        added_input_connectors,
+        modified_input_connectors,
+        removed_input_connectors,
+        added_output_connectors,
+        modified_output_connectors,
+        removed_output_connectors,
+    })
+}
+
+/// Reasons bootstrapping cannot be performed.
+struct BootstrapBlockers {
```

Comment:
we actually allow lateness annotations on views too, in the middle of a pipeline. I wonder what happens about these. They can be even local views.